draco/synthesis/mockcatalog.py

```diff
@@ -0,0 +1,1045 @@
+
+    def process(self, source_map):
+        """Make PDF map that assumes a trivial selection function.
+
+        Parameters
+        ----------
+        source_map : :class:`containers.Map`
+            Overdensity map that determines z and angular resolution
+            of output PDF map.
+
+        Returns
+        -------
+        pdf_map : :class:`containers.Map`
+            Output PDF map.
+        """
+
+        # Get local offset and shape of frequency axis, and global shape
+        lo = source_map.map.local_offset[0]
+        ls = source_map.map.local_shape[0]
+        gs = source_map.map.global_shape[0]
+
+        if not self.use_voxel_volumes:
+            # Set each frequency channel to have equal total probability
+            z_weights = mpiarray.MPIArray.wrap(1 / gs * np.ones(ls), axis=0)
+
+        else:
+            # Set total probability for each frequency channel based
+            # on voxel volume for that channel.
+            # Healpix maps have equal-angular-area pixels, so the voxel
+            # area is proportional to \chi^2 * (\chi_max - \chi_min),
+            # where we use \chi_centre for the first \chi (which incorporates
+            # the z-dependence of transverse area), and the second factor
+            # is the voxel size along the z direction.
+            from cora.util import cosmology
+
+            cosmo = cosmology.Cosmology()
+            z_weights_global = np.zeros(gs, dtype=np.float64)
+
+            # First, we compute the normalization for each channel
+            # globally
+            for fi, freq in enumerate(source_map.index_map["freq"]):
+                z_min = units.nu21 / (freq[0] + 0.5 * freq[1]) - 1
+                z_max = units.nu21 / (freq[0] - 0.5 * freq[1]) - 1
+                z_mean = units.nu21 / freq[0] - 1
+
+                z_weights_global[fi] = cosmo.comoving_distance(z_mean) ** 2 * (
+                    cosmo.comoving_distance(z_max) - cosmo.comoving_distance(z_min)
+                )
+
+            z_weights_global /= z_weights_global.sum()
+
+            # Select local section of weights
+            z_weights = mpiarray.MPIArray.wrap(z_weights_global[lo : lo + ls], axis=0)
+
+        # Create PDF map
+        pdf_map = self.make_pdf_map(source_map, z_weights)
+
+        return pdf_map
+
+
+class MockCatGenerator(task.SingleTask):
+    """Take PDF maps generated by task :class:`PdfGenerator`
+    and use it to draw mock catalogs.
+
+    Attributes
+    ----------
+    nsources : int
+        Number of sources to draw in each mock catalog.
+    ncats : int
+        Number of catalogs to generate.
+    sigma_z : float, optional
+        Standard deviation of Gaussian redshift errors (default: None)
+    sigma_z_over_1plusz : float, optional
+        Standard deviation of Gaussian redshift errors will be set to
+        this parameter times (1+z). Only one of this and `sigma_z` can
+        be specified. Default: None
```

Comment:
Should the redshift error generation be a separate task? I can see it being useful elsewhere, e.g. maybe you want to artificially degrade a real catalog.

Reply:
Okay, I split it into another task. I'd like to add the option for the user to input a custom PDF, but in the interest of getting this merged quickly, maybe I'll leave that for a separate PR